index.bs

```diff
@@ -1289,52 +1309,101 @@ SessionResult = (StatusResult)
+
+             1. Set |enabled events|[|event name|] to |context enabled
+                events|.
+
+      1. If |enabled| is false:
+
+        1. For each |event name| in |event names|:
+
+          1. If the |global event set| [=contains=] |event name|, remove |event
+             name| from the |global event set|.
+
+      1. Return [=success=] with data |enabled events|.
+
+  1. Assert: |browsing contexts| is not null.
 
-      1. If |enabled| is true, for each |event name| in |event names|,
-         append |event name| to |global event set|. Otherwise for for each
-         |event name| in |event names|, if the |global event set| [=contains=]
-         |event name|, remove |event name| from the |global event set|.
+  1. Let |targets| be an empty map.
 
-      1. Return
+  1. For each |context id| in |browsing contexts|:
 
-    1. Let |targets| be an empty list.
+    1. Let |context| be the result of [=trying=] to [=get a browsing context=]
+       with |context id|.
+
+    1. Let |top-level context| be the [=top-level browsing context=] for |context|.
+
+    1. If |event map| does not contain |top-level context|, set |event
+       map|[|top-level context|] to a new set.
+
+    1. Set |targets|[|top-level context|] to |event map|[|top-level context|].
+
+  1. For each |event name| in |event names|:
+
+    1. If |enabled| is true and |global event set| contains |event name|, continue.
 
-    1. For each entry |context id| in the |list of contexts|:
+    1. For each |context| → |target| in |targets|:
 
-      1. Let |context| be the result of [=trying=] to [=get a browsing context=]
-         with |context id|.  If the |event map| does not contain an entry for
-         |context|, set the value of the entry for |context| to a new empty map.
+      1. If |enabled| is true and |target| does not contain |event name|:
 
-      1. Get the entry from the |event map| for |context| and append it to
-         |targets|.
+         1. Append |event name| to |target|.
 
-    1. For each |target| in |targets|:
+         1. If |enabled events| does not contain |event name|, set |enabled
+            events|[|event name|] to a new set.
 
-      1. For each |event name| in |event names|:
+         1. Append |context| to |context enabled events|[|event name|].
 
-        1. Set |target|[|event name|] to |enabled|.
+      1. If |enabled| is false and |target| contains |event name|, remove |event
+         name| from |target|.
 
-    1. Return [=success=] with data null.
+  1. Return [=success=] with data |enabled events|.
+
+Note: Implementations that do additional work when an event is enabled,
+e.g. subscribing to the relevant engine-internal events, will likely perform
+those additional steps when updating the event map. This specification uses
+a model where hooks are always called and then the event map is used to
+filter only those that ought to be returned to the local end.
+
+Issue: Currently if an event is enabled globally and then disabled for a
```

Comment:
Should be an error unless we have a disabled event context map as well.

Reply:
OK, so I updated this to emit an error in the case that you try to disable an event that is not in fact enabled. idk if this is the right call here, or if we should just do this for the global case, but not the context-specific case.